src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+    switch (ip4Header.GetProtocol())
+    {
+    case Ip4::kProtoUdp:
+        ip6Header.SetNextHeader(Ip6::kProtoUdp);
+        res = Result::kForward;
+        break;
+    case Ip4::kProtoTcp:
+        ip6Header.SetNextHeader(Ip6::kProtoTcp);
+        res = Result::kForward;
+        break;
+    case Ip4::kProtoIcmp:
+        ip6Header.SetNextHeader(Ip6::kProtoIcmp6);
+        SuccessOrExit(TranslateIcmp4(*mapping, aMessage));
+        res = Result::kForward;
+        break;
+    default:
+        ExitNow(res = Result::kDrop);
+    }
+
+    switch (res)
+    {
+    case Result::kDrop:
+        break;
+    case Result::kReplyIcmp:
+        // TODO: Implement the logic for replying ICMP packets.
+        break;
+    case Result::kForward:
+        ip6Header.SetPayloadLength(aMessage.GetLength() - aMessage.GetOffset());
+        Checksum::UpdateMessageChecksum(aMessage, ip6Header.GetSource(), ip6Header.GetDestination(),
+                                        ip6Header.GetNextHeader());
+        err = aMessage.PrependBytes(&ip6Header, sizeof(ip6Header));
+        if (err != kErrorNone)
+        {
+            // This might happen when the platform failed to reserve enough space before the original IPv4 packet.
+            LogWarn("failed to prepend IPv6 head to translated message");
+            res = Result::kDrop;
+        }
+        break;
+    }
+
+exit:
+    return res;
+}
+
+void Nat64Translator::ReleaseMapping(AddressMapping &aMapping)
+{
+    LogInfo("mapping removed: %s -> %s", aMapping.mIp6.ToString().AsCString(), aMapping.mIp4.ToString().AsCString());
+    mIp4AddressPool[mAvailableAddressCount] = aMapping.mIp4;
+    mAvailableAddressCount++;
+    mAddressMappingPool.Free(aMapping);
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::CreateMapping(const Ip6::Address &aAddr)
+{
+    AddressMapping *mapping = mAddressMappingPool.Allocate();
+
+    if (mapping == nullptr)
+    {
+        uint64_t                   now = Get<Uptime>().GetUptime();
+        LinkedList<AddressMapping> idleMappings;
+
+        mActiveAddressMappings.RemoveAllMatching(now, idleMappings);
+
+        for (AddressMapping *idleMapping = idleMappings.Pop(); idleMapping != nullptr; idleMapping = idleMappings.Pop())
+        {
+            ReleaseMapping(*idleMapping);
+        }
+
+        mapping = mAddressMappingPool.Allocate();
+    }
+
+    if (mAvailableAddressCount == 0)
+    {
+        if (mapping != nullptr)
+        {
+            mAddressMappingPool.Free(*mapping);
+            mapping = nullptr;
+        }
+        ExitNow();
+    }
+
+    VerifyOrExit(mapping != nullptr);
+
+    mapping->mIp6 = aAddr;
+    mAvailableAddressCount--;
+    mapping->mIp4 = mIp4AddressPool[mAvailableAddressCount];
+    mapping->Touch(Get<Uptime>().GetUptime());
+    mActiveAddressMappings.Push(*mapping);
+    LogInfo("mapping created: %s -> %s", mapping->mIp6.ToString().AsCString(), mapping->mIp4.ToString().AsCString());
```

Comment:
Can consider adding a `ToString()` method to `AddressMapping` class which returns a `String` containing the  `"%s -> %s"` with the addresses.